### PR TITLE
feat(eslint-plugin): add rule object-curly-spacing

### LINKED ISSUE #1784: [object-curly-spacing] One-line object types do not adhere to spacing rules
**Repro**

```JSON
{
  "rules": {
    "object-curly-spacing": "never"
  }
}
```

```TS
const foo = ({ status }: { status: int }) => console.log(status);
```

**Expected Result**
That running `eslint --fix` would fix this to:
```TS
const foo = ({status}: {status: int}) => console.log(status);
```

**Actual Result**
Instead, only the first curly brackets are fixed:
```TS
const foo = ({status}: { status: int }) => console.log(status);
```

**Additional Info**
This rule does not exist yet in `@typescript-eslint`, but should be extended from `eslint` like many other rules (e.g. [`brace-style`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/brace-style.md))

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.24.0` |
| `@typescript-eslint/parser`        | `2.24.0` |
| `TypeScript`                       | `3.6.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `13.11.0` |
| `npm`                              | `6.13.7` |

### PR DESCRIPTION
**Pull request by Kiikurage**
added new extended rule `object-curly-spacing`

- [x] tests
- [x] docs
- [x] typing update

Fixes #1784

### CODE REVIEW

**Comment by mpsijm** [comment]
@bradzacher Wat would need to happen before this PR can be reviewed (besides perhaps @Kiikurage rebasing the branch on master because of the merge conflict)?

**Comment by bradzacher** [comment]
https://github.com/typescript-eslint/typescript-eslint/blob/master/CONTRIBUTING.md#addressing-feedback-and-beyond

File: packages/eslint-plugin/src/rules/object-curly-spacing.ts [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+      unexpectedSpaceBefore: "There should be no space before '{{token}}'.",
+      unexpectedSpaceAfter: "There should be no space after '{{token}}'.",
+    },
+    fixable: baseRule.meta.fixable,
+    schema: baseRule.meta.schema,
+  },
+  defaultOptions: ['never'],
+  create(context) {
+    const spaced = context.options[0] === 'always',
+      sourceCode = context.getSourceCode();
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/object-curly-spacing.ts:41]: nit
```suggestion
    const spaced = context.options[0] === 'always';
    const sourceCode = context.getSourceCode();
```

File: packages/eslint-plugin/src/rules/object-curly-spacing.ts [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+    /**
+     * Determines whether an option is set, relative to the spacing option.
+     * If spaced is "always", then check whether option is set to false.
+     * If spaced is "never", then check whether option is set to true.
+     * @param {Object} option The option to exclude.
+     * @returns {boolean} Whether or not the property is excluded.
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/object-curly-spacing.ts:48]: nitpick

```suggestion
     * @param option The option to exclude.
     * @returns Whether or not the property is excluded.
```

File: packages/eslint-plugin/src/rules/object-curly-spacing.ts [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+      ...rules,
+      TSTypeLiteral(node: TSESTree.TSTypeLiteral): void {
+        if (node.members.length === 0) {
+          return;
+        }
+
+        const first = sourceCode.getFirstToken(node)!,
+          last = getClosingBraceOfObject(node)!,
+          second = sourceCode.getTokenAfter(first, { includeComments: true })!,
+          penultimate = sourceCode.getTokenBefore(last, {
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/object-curly-spacing.ts:265]: nit
```suggestion
        const first = sourceCode.getFirstToken(node)!;
        const last = getClosingBraceOfObject(node)!;
        const second = sourceCode.getTokenAfter(first, { includeComments: true })!;
        const penultimate = sourceCode.getTokenBefore(last, {
```

**Pull request closed without merging**